src/haddock/libs/libgrid.py

```diff
@@ -395,119 +393,401 @@ def __repr__(self) -> str:
-                #  seconds, so using multiple threads can speed up the process
-                #
-                # `ThreadPoolExecutor` is used to initialize a pool of threads, and
-                #  then `executor.map` is used to apply the `process_job` method to each
-                #  job, this will effectively check the status of all jobs in parallel,
-                #  using `self.params["ncores"]` threads.
-                # =====================================================================#
-                with ThreadPoolExecutor(max_workers=self.params["ncores"]) as executor:
-                    # =====================================================================#
-                    # EXPLANATION #
-                    # =====================================================================#
-                    # The function `process_job` will return a tuple with the job and a
-                    #  boolean. This return values are captured in `results` list.
-                    #  The boolean will be True if the job is done (either successfully or
-                    #  failed after retries), and False otherwise - meaning the job is
-                    #  still running or waiting.
-                    # =====================================================================#
-                    results = list(executor.map(self.process_job, pending_jobs))
-
-                for job, is_done in results:
-                    # =====================================================================#
-                    # EXPLANATION #
-                    # =====================================================================#
-                    # Here we update the `queue` dictionary to avoid checking the jobs that
-                    #  are already done.
-                    # =====================================================================#
-                    if is_done:
-                        queue[job] = True
-
-            # Do some simple logging to keep track of progress.
-            done = sum(1 for done in queue.values() if done)
-            log.info(f"{done}/{total} jobs completed.")
-            complete = all(queue.values())
-
-        log.info("All jobs completed.")
+            # Calculate running duration (RUNNING to DONE)
+            running_duration = (done_time - run_start).total_seconds()
+            running_durations.append(running_duration)
+
+        if not running_durations or not waiting_durations:
+            log.warning(
+                "> Average running time is zero, cannot calculate optimal batch size <"
+            )
+            return
+
+        # Calculate average durations
+        avg_waiting = sum(waiting_durations) / len(waiting_durations)
+        avg_running = sum(running_durations) / len(running_durations)
+
+        target_efficiency = 0.9
+        batch_size = self.calculate_optimal_batch_size(
+            N=1,  # each batch had one job
```

Comment:
What about using the actual value that was used during the probing ?
```suggestion
            N=self.batch_size,
```

Reply:
Yep, that also works